src/py/flwr/common/backoff.py

```diff
@@ -0,0 +1,278 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""`SafeInvoker` to argument other functions with error handling and retries."""
+
+
+import itertools
+import random
+import time
+from dataclasses import dataclass
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    Optional,
+    Tuple,
+    Type,
+    Union,
+)
+
+
+def exponential(
+    base_delay: float = 1,
+    multiplier: float = 2,
+    max_delay: Optional[int] = None,
+) -> Generator[float, None, None]:
+    """Wait time generator for exponential backoff strategy.
+
+    Parameters
+    ----------
+    base_delay: float (default: 1)
+        Initial delay duration before the first retry.
+    multiplier: float (default: 2)
+        Factor by which the delay is multiplied after each retry.
+    max_delay: Optional[float] (default: None)
+        The maximum delay duration between two consecutive retries.
+    """
+    delay = base_delay if max_delay is None else min(base_delay, max_delay)
+    while True:
+        yield delay
+        delay *= multiplier
+        if max_delay is not None:
+            delay = min(delay, max_delay)
+
+
+def constant(
+    interval: Union[float, Iterable[float]] = 1,
+) -> Generator[float, None, None]:
+    """Wait time generator for specified intervals.
+
+    Parameters
+    ----------
+    interval: Union[float, Iterable[float]] (default: 1)
+        A constant value to yield or an iterable of such values.
+    """
+    if not isinstance(interval, Iterable):
+        interval = itertools.repeat(interval)
+    yield from interval
+
+
+def random_jitter(value: float) -> float:
+    """Jitter the value by a random number of milliseconds.
+
+    This adds up to 1 second of additional time to the original value.
+    Prior to backoff version 1.2, this was the default jitter behavior.
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return value + random.random()
+
+
+def full_jitter(value: float) -> float:
+    """Jitter the value across the full range (0 to value).
+
+    This corresponds to the "Full Jitter" algorithm specified in the
+    AWS blog post on the performance of various jitter algorithms.
+    See: http://www.awsarchitectureblog.com/2015/03/backoff.html
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return random.uniform(0, value)
+
+
+@dataclass
+class Details:
```

Comment:
Depending on how we name the class below, we should align the name here as well. Maybe a `RetryCallable`?

Reply:
sry but I don't quite follow. Do you mean to change the class name `Details` to `RetryCallable`?

Reply:
Yes, that was the idea. Or maybe just `Retryable`?

Reply:
As discussed earlier, I've renamed `Details` to `RetryState`.